### ISSUE TITLE
[QST] Is this the right way to set up a cuda cluster on a machine with a single GPU?

### ISSUE THREAD

**Issue by Arnie0426** [NVIDIA/NVTabular#1000]
After seeing a couple of my jobs run OOM on a single GPU instance, I came across this: https://nvidia.github.io/NVTabular/main/resources/troubleshooting.html#initializing-a-dask-cuda-cluster

But when I set a local dask cuda cluster, the workflow just seems to hang for a long time. Am I initializing something incorrectly? Is a Local CUDA Cluster absolutely necessary? Is there any other way to avoid OOM errors?

```python
# External dependencies
import os
import cudf 
import shutil
import numpy as np

import nvtabular as nvt

from dask_cuda import LocalCUDACluster
import torch
from dask.distributed import Client
from nvtabular.utils import device_mem_size

# Use total device size to calculate args.device_limit_frac
device_size = device_mem_size(kind="total")

part_mem_frac = 0.8
device_limit_frac = 0.5  # Spill GPU-Worker memory to host at this limit.
device_pool_frac = 0.15

device_limit = int(device_limit_frac * device_size)
dask_workdir = "/usr/local/mnt/data/dask_workdir"
part_size = int(part_mem_frac * device_size)
device_pool_size = int(device_pool_frac * device_size)

cluster = LocalCUDACluster(
    n_workers=torch.cuda.device_count(),                        # Number of GPU workers
    device_memory_limit=device_limit,         # GPU->CPU spill threshold (~75% of GPU memory)
    rmm_pool_size=device_limit * 1.1,               # Memory pool size on each worker
    local_directory=dask_workdir,   # Fast directory for disk spilling
)

client = Client(cluster)

train_set = nvt.Dataset(
                data_path,
                engine="parquet",
                part_size="100MB",
            )

grouped = ["author", "product_id"] >> nvt.ops.Groupby(
                groupby_cols=["author"], aggs="count"
            )
workflow = nvt.Workflow(grouped, client=client)
grouped_users: nvt.Dataset = workflow.fit_transform(train_set)

```

It just hangs when fit_transform() is called on the train_set and I don't see any GPU utilization.

**Comment by Arnie0426** [comment]
For some background, this "train_set" is around 50 GB in size, and if I don't use a LocalCUDACluster, it silently OOMs on me

**Comment by rjzamora** [comment]
Thanks for sharing your workflow @Arnie0426 - Could you say a bit more about the machine (e.g. what is the single-device memory, and how many devices are there) and the dataset.  Regarding the dataset: How many files are there, and how big are the files and row-groups?  Can you share what `train_set.validate_dataset()` returns?

**Comment by Arnie0426** [comment]
Absolutely, it's a GPU pod running on EKS (p3.2xlarge) with 50 gigs of memory and 6 cpus.

I learned that I am also running OOM on a smaller dataset.

Dataset: There are a bit over 2000 parquet files on S3 - each around 2MB in size. (~2.5 GB total)

I just ran `validate_dataset()` on it (for the first time), and it returned False and suggested that I regenerate the dataset with larger files - (which makes sense, but is a bit hard for us to do upstream, given that these files are generated by a spark cluster with lots of partitions, and `coalesce`ing has its own costs)

Is regenerating the data with `regenerate_dataset()` with larger files essential for this?

**Comment by rjzamora** [comment]
>Dataset: There are a bit over 2000 parquet files on S3 - each around 2MB in size. (~2.5 GB total)

Ah - Interesting.  If your dataset does not have a `_metadata` file, then the initial `Dataset()` metadata-collection stage may be very expensive (resulting in the apparent "hang").  For example, I am wondering if this also hangs: `import dask_cudf; dask_cudf.read_parquet(data_path)`

>Is regenerating the data with regenerate_dataset() with larger files essential for this?

Hopefully not.  However, you *may* need to add a `_metadata` file if you don't already have one.

**Comment by Arnie0426** [comment]
Yeah, that must be it. Is there a suggested way to generate that `_metadata` file? Spark doesn't seem to create it by default anymore.

**Comment by rjzamora** [comment]
Your best bet may be [Dask's `create_metadata_file` utility](https://github.com/dask/dask/blob/5fbbf90a59c19c197875345fb215d536fc451b8d/dask/dataframe/io/parquet/core.py#L725), or to use `train_set.validate_dataset(add_metadata_file=True)`
